### PR TITLE
⬇🗄always restore one url to one place

### PR DESCRIPTION
**Pull request by OsmondJiang**
1. always return one ulr to one place
2. remove etag from file name
3. add process lock when reading/writing restored file

will rename RestoreMap later

### CODE REVIEW

File: src/docfx/restore/RestoreFile.cs [hunk 1]
```diff
@@ -114,6 +68,14 @@ public static IEnumerable<string> GetFileReferences(this Config config)
             yield return config.MonikerDefinition;
         }
 
+        public static string GetRestorePath(string url)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(url));
+            Debug.Assert(HrefUtility.IsHttpHref(url));
+
+            return PathUtility.NormalizeFile(Path.Combine(AppData.GetFileDownloadDir(url), HashUtility.GetMd5HashShort(url)));
```
Comment (yufeih) [inline src/docfx/restore/RestoreFile.cs]: `GetFileDownloadDir` already contains url hash, so no need for `HashUtility.GetMd5HashShort(url)`

File: docs/designs/restore-and-gc.md [hunk 2]
```diff
@@ -20,58 +20,39 @@ github:
...
+And there is another etag file `{url-hash}.etag` under `%DOCFX_APPDATA_PATH%/downloads/{url-short-path}/` tracking the `etag` info of this file if has.
 
 ```text
 `%DOCFX_APPDATA_PATH%`
     | - downloads
         | - raw.gith..tent.com+docascode+docfx-te..ndencies+62b0448+extend1.yml+dc363b0e
-            | - 1
-            | - 2
-            | - index.json
+            | - dc363b0e
```
Comment (superyyrrzz) [inline docs/designs/restore-and-gc.md]: `dc363b0e` is already part of the parent directory name. Can use `content` and `etag` as filename.

File: docs/designs/restore-and-gc.md [hunk 3]
```diff
@@ -20,58 +20,39 @@ github:
...
-- `{date}` is the last access date, it will be set or overwritten during restore.
 
-We are using [Shared/Exclusive Lock](https://www.ibm.com/support/knowledgecenter/en/SSGMCP_5.1.0/com.ibm.cics.ts.applicationprogramming.doc/topics/dfhp39o.html) to trakcing each index using states:
+Same url will always be restored to the same place, and we are using `Process Lock` to read/write the file and its etag file.
 
-- During build, we acquire an available index from pool with `shared lock`.
-- During restore, we acquire an available index from pool with `exclusive lock`.
-
->NOTE: Shared and Exclusived Lock need to be cross process, we are not going to discuss about that details here
+> NOTE: docfx will load all restored files' content at the first stage of build, never load again, in case ths file is restored again(changed) in current build.
```
Comment (superyyrrzz) [inline docs/designs/restore-and-gc.md]: typo: ths

File: docs/designs/restore-and-gc.md [hunk 4]
```diff
@@ -20,58 +20,39 @@ github:
...
-The restored files will be stored at `%DOCFX_APPDATA_PATH%/downloads/{url-short-path}/{number}` like:
+The restored files will be stored at `%DOCFX_APPDATA_PATH%/downloads/{url-short-path}/{url-hash}` like:
 
 ```text
 `%DOCFX_APPDATA_PATH%`
     | - downloads
         | - raw.gith..tent.com+docascode+docfx-te..ndencies+62b0448+extend1.yml+dc363b0e
-            | - 1
-            | - 2
+            | - dc363b0e
```
Comment (yufeih) [inline docs/designs/restore-and-gc.md]:
```suggestion
        | - raw.gith..tent.com+docascode+docfx-te..ndencies+62b0448+extend1.yml+dc363b0e.etag
```

File: src/docfx/restore/RestoreFile.cs (hunk 1 above)
Comment (OsmondJiang) [reply src/docfx/restore/RestoreFile.cs]: AppData.GetFileDownloadDir(url) is the dir of this file, I like @superyyrrzz 's idea, `content` and `etag`

**Pull request merged**